## Email Similarity

In this project, you will use scikit-learn’s **Naive Bayes** implementation on several different datasets. By reporting the accuracy of the classifier, we can find which datasets are harder to distinguish. For example, how difficult do you think it is to distinguish the difference between emails about hockey and emails about soccer? How hard is it to tell the difference between emails about hockey and emails about tech? In this project, we’ll find out exactly how difficult those two tasks are.

In [1]:
from sklearn.datasets import fetch_20newsgroups
from sklearn.naive_bayes import MultinomialNB
from sklearn.feature_extraction.text import CountVectorizer

emails = fetch_20newsgroups()

1. We’ve imported a dataset of emails from scikit-learn’s datasets. All of these emails are tagged based on their content.

    Print `emails.target_names` to see the different categories.


## Exploring the Data

In [2]:
print(emails.target_names)

['alt.atheism', 'comp.graphics', 'comp.os.ms-windows.misc', 'comp.sys.ibm.pc.hardware', 'comp.sys.mac.hardware', 'comp.windows.x', 'misc.forsale', 'rec.autos', 'rec.motorcycles', 'rec.sport.baseball', 'rec.sport.hockey', 'sci.crypt', 'sci.electronics', 'sci.med', 'sci.space', 'soc.religion.christian', 'talk.politics.guns', 'talk.politics.mideast', 'talk.politics.misc', 'talk.religion.misc']


2. We’re interested in seeing how effective our Naive Bayes classifier is at telling the difference between a baseball email and a hockey email. We can select the categories of articles we want from `fetch_20newsgroups` by adding the parameter `categories`.

    In the function call, set `categories` equal to the list `['rec.sport.baseball', 'rec.sport.hockey']`


In [3]:
emails = fetch_20newsgroups(categories = ['rec.sport.baseball', 'rec.sport.hockey'])

3. Let’s take a look at one of these emails.

    All of the emails are stored in a list called `emails.data`. Print the email at index `5` in the list.


In [4]:
print(emails.data[5])

From: mmb@lamar.ColoState.EDU (Michael Burger)
Subject: More TV Info
Distribution: na
Nntp-Posting-Host: lamar.acns.colostate.edu
Organization: Colorado State University, Fort Collins, CO  80523
Lines: 36

United States Coverage:
Sunday April 18
  N.J./N.Y.I. at Pittsburgh - 1:00 EDT to Eastern Time Zone
  ABC - Gary Thorne and Bill Clement

  St. Louis at Chicago - 12:00 CDT and 11:00 MDT - to Central/Mountain Zones
  ABC - Mike Emerick and Jim Schoenfeld

  Los Angeles at Calgary - 12:00 PDT and 11:00 ADT - to Pacific/Alaskan Zones
  ABC - Al Michaels and John Davidson

Tuesday, April 20
  N.J./N.Y.I. at Pittsburgh - 7:30 EDT Nationwide
  ESPN - Gary Thorne and Bill Clement

Thursday, April 22 and Saturday April 24
  To Be Announced - 7:30 EDT Nationwide
  ESPN - To Be Announced


Canadian Coverage:

Sunday, April 18
  Buffalo at Boston - 7:30 EDT Nationwide
  TSN - ???

Tuesday, April 20
  N.J.D./N.Y. at Pittsburgh - 7:30 EDT Nationwide
  TSN - ???

Wednesday, April 21
  St. Louis a

4. All of the labels can be found in the list `emails.target`. Print the label of the email at index `5`.

    The labels themselves are numbers, but those numbers correspond to the label names found at `emails.target_names`.

    Is this a baseball email or a hockey email?

In [5]:
print(emails.target[5])
print(emails.target_names[1])

1
rec.sport.hockey


## Making the Training and Test Sets

5. We now want to split our data into training and test sets. Change the name of your variable from `emails` to `train_emails`. Add these three parameters to the function call:

    - `subset='train'`
    - `shuffle = True`
    - `random_state = 108`

Adding the `random_state` parameter will make sure that every time you run the code, your dataset is split in the same way.


In [6]:
train_emails = fetch_20newsgroups(categories = ['rec.sport.baseball', 'rec.sport.hockey'], subset = 'train', shuffle = True, random_state = 108)

6. Create another variable named `test_emails` and set it equal to `fetch_20newsgroups`. The parameters of the function should be the same as before except `subset` should now be `'test'`.

In [7]:
test_emails = fetch_20newsgroups(categories = ['rec.sport.baseball', 'rec.sport.hockey'], subset = 'test', shuffle = True, random_state = 108)

## Counting Words

7. We want to transform these emails into lists of word counts. The `CountVectorizer` class makes this easy for us.

    Create a `CountVectorizer` object and name it `counter`.


In [8]:
counter = CountVectorizer()

8. We need to tell counter what possible words can exist in our emails. counter has a `.fit()` a function that takes a list of all your data.

    Call `.fit()` with `test_emails.data + train_emails.data` as a parameter.


In [9]:
counter.fit(test_emails.data + train_emails.data)

CountVectorizer()

8. We can now make a list of the counts of our words in our training set.

    Create a variable named `train_counts`. Set it equal to `counter`‘s `transform` function using `train_emails.data` as a parameter.


In [10]:
train_counts = counter.transform(train_emails.data)

10. Let’s also make a variable named `test_counts`. This should be the same function call as before, but use `test_emails.data` as the parameter of `transform`.


In [11]:
test_counts = counter.transform(test_emails.data)

## Making a Naive Bayes Classifier

11. Let’s now make a Naive Bayes classifier that we can train and test on. Create a `MultinomialNB` object named `classifier`.

In [12]:
classifier = MultinomialNB()

12. Call `classifier`‘s `.fit()` function. `.fit()` takes two parameters. The first should be our training set, which for us is `train_counts`. The second should be the labels associated with the training emails. Those are found in `train_emails.target`.

In [13]:
classifier.fit(train_counts, train_emails.target)

MultinomialNB()

13. Test the Naive Bayes Classifier by printing classifier‘s `.score()` function. `.score()` takes the test set and the test labels as parameters.

    `.score()` returns the accuracy of the classifier on the test data. Accuracy measures the percentage of classifications a classifier correctly made.


In [14]:
classifier.score(test_counts, test_emails.target)

0.9723618090452262

## Testing Other Datasets

14. Our classifier does a pretty good job distinguishing between soccer emails and hockey emails. But let’s see how it does with emails about really different topics.

    Find where you created `train_emails` and `test_emails`. Change the categories to be `['comp.sys.ibm.pc.hardware','rec.sport.hockey']`.

    Did your classifier do a better or worse job on these two datasets?


In [15]:
train_emails = fetch_20newsgroups(categories = ['comp.sys.ibm.pc.hardware','rec.sport.hockey'], subset = 'train', shuffle = True, random_state = 108)

test_emails = fetch_20newsgroups(categories = ['comp.sys.ibm.pc.hardware','rec.sport.hockey'], subset = 'test', shuffle = True, random_state = 108)

counter = CountVectorizer()
counter.fit(test_emails.data + train_emails.data)

train_counts = counter.transform(train_emails.data)

test_counts = counter.transform(test_emails.data)

classifier = MultinomialNB()
classifier.fit(train_counts, train_emails.target)
print(classifier.score(test_counts, test_emails.target))


0.9974715549936789


*It does a better job.*

15. Play around with different sets of data. Can you find a set that’s incredibly accurate or incredibly inaccurate?

The possible categories are listed below.

-    `'alt.atheism'`
-    `'comp.graphics'`
-    `'comp.os.ms-windows.misc'`
-    `'comp.sys.ibm.pc.hardware'`
-    `'comp.sys.mac.hardware'`
-    `'comp.windows.x'`
-    `'misc.forsale'`
-    `'rec.autos'`
-    `'rec.motorcycles'`
-    `'rec.sport.baseball'`
-    `'rec.sport.hockey'`
-    `'sci.crypt'`
-    `'sci.electronics'`
-    `'sci.med'`
-    `'sci.space'`
-    `'soc.religion.christian'`
-    `'talk.politics.guns'`
-    `'talk.politics.mideast'`
-    `'talk.politics.misc'`
-    `'talk.religion.misc'`


In [16]:
train_emails = fetch_20newsgroups(categories = ['comp.windows.x','comp.sys.ibm.pc.hardware'], subset = 'train', shuffle = True, random_state = 108)

test_emails = fetch_20newsgroups(categories = ['comp.windows.x','comp.sys.ibm.pc.hardware'], subset = 'test', shuffle = True, random_state = 108)

counter = CountVectorizer()
counter.fit(test_emails.data + train_emails.data)

train_counts = counter.transform(train_emails.data)

test_counts = counter.transform(test_emails.data)

classifier = MultinomialNB()
classifier.fit(train_counts, train_emails.target)
print(classifier.score(test_counts, test_emails.target))


0.9428208386277002


- `['rec.autos','rec.motorcycles']` = 97%
- `['soc.religion.christian','talk.politics.guns']` = 99%
- `['comp.windows.x','talk.politics.mideast']` = 99%
- `['comp.windows.x','comp.sys.ibm.pc.hardware']` = 94%